### PR TITLE
feat(api-domains-auth-admin): Multi env admin api with merging

### PR DESCRIPTION
**Pull request by saevarma**
# Multi env admin api with merging

[Notion task](https://www.notion.so/Merge-domains-in-GraphQL-b353ce27833045ff861f89f19463f958)

## What

Add client library for Admin Api and setup multi env apis for the auth admin domain to query data for dev, staging and prod then merge the results.

## Why

To provide data connection for the new IDS Admin module.

## Screenshots / Gifs

N/A

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Formatting passes locally with my changes
- [ ] I have rebased against main before asking for a review

### CODE REVIEW

File: apps/api/infra/api.ts [hunk 1]
```diff
@@ -181,6 +181,18 @@ export const serviceSetup = (services: {
...
+      AUTH_ADMIN_API_URL_STAGING: {
+        dev: '',
+        staging: 'https://identity-server-admin.staging01.devland.is',
+        prod: 'https://identity-server-admin.staging01.devland.is',
+      },
+      AUTH_ADMIN_API_URL_PROD: {
+        dev: '',
+        staging: '',
+        prod: 'https://identity-server-admin.innskra.island.is',
+      },
```
Comment (saevarma) [inline apps/api/infra/api.ts:195]: @island-is/devops  I need help here to configure the URLs and then we need to do some AWS magic to query cross environments 😅

cc @busla

Comment (eirikurn) [reply apps/api/infra/api.ts:195]:
```
DEV: https://services-auth-admin-api.internal.dev01.devland.is
STAGING: https://services-auth-admin-api.internal.staging01.devland.is
PROD: https://services-auth-admin-api.internal.innskra.island.is
```

Let's go with this until we know more.

File: libs/clients/auth/admin-api/README.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+Which can be used in a service like:
+
+```typescript
+import { AdminDevApi } from '@island.is/auth-admin-api'
+
+@Injectable()
+export class MyService {
+  constructor(
+    @Inject(AdminDevApi.key)
+    @Optional()
```
Comment (eirikurn) [inline libs/clients/auth/admin-api/README.md:28]: Is `@Optional()` needed if the factory returns `undefined`?

**Review by eirikurn — APPROVED** [review]
Nice, let's get this in and improve in later PRs 👍

File: libs/clients/auth/admin-api/README.md (hunk 2 above)
Comment (saevarma) [reply libs/clients/auth/admin-api/README.md:28]: Nope. Removed it.

File: libs/api/domains/auth-admin/src/lib/tenant/tenants.service.ts [hunk 3]
```diff
@@ -1,112 +1,92 @@
...
+    return this.adminDevApi?.withMiddleware(new AuthMiddleware(auth))
+  }
+  private adminStagingApiWithAuth(auth: Auth) {
+    return this.adminStagingApi?.withMiddleware(new AuthMiddleware(auth))
+  }
+  private adminProdApiWithAuth(auth: Auth) {
+    return this.adminProdApi?.withMiddleware(new AuthMiddleware(auth))
+  }
+
+  async getTenants(user: User): Promise<TenantsPayload> {
```
[the remark this answers is not in the document]
Comment (saevarma) [reply libs/api/domains/auth-admin/src/lib/tenant/tenants.service.ts:50]: Yes we will do that in follow up PRs when adding applications.

File: apps/api/infra/api.ts [hunk 4]
```diff
@@ -181,6 +181,22 @@ export const serviceSetup = (services: {
...
+          staging:
+            'https://services-auth-admin-api.internal.staging01.devland.is',
+        }),
+        prod: json({
+          dev: 'https://services-auth-admin-api.internal.dev01.devland.is',
+          staging:
+            'https://services-auth-admin-api.internal.staging01.devland.is',
+          prod: 'https://services-auth-admin-api.internal.innskra.island.is',
+        }),
+      },
```
Comment (saevarma) [inline apps/api/infra/api.ts:199]: @eirikurn using JSON instead?
Only a bit awkward defining `dev` and `staging` path multiple times 🤔

File: apps/api/infra/api.ts [hunk 5]
```diff
@@ -181,6 +181,22 @@ export const serviceSetup = (services: {
...
+          dev: 'https://services-auth-admin-api.internal.dev01.devland.is',
+          staging:
+            'https://services-auth-admin-api.internal.staging01.devland.is',
+        }),
+        prod: json({
+          dev: 'https://services-auth-admin-api.internal.dev01.devland.is',
+          staging:
+            'https://services-auth-admin-api.internal.staging01.devland.is',
+          prod: 'https://services-auth-admin-api.internal.innskra.island.is',
+        }),
```
Comment (eirikurn) [inline apps/api/infra/api.ts:198]: That's fine by me. Note that dev and staging use intra-account communications. So maybe something like this (if it works).

```suggestion
        dev: ref((h) => json({
          dev: `http://${h.svc(services.authAdminApi)}`,
        })),
        staging: ref((h) => json({
          dev: 'https://services-auth-admin-api.internal.dev01.devland.is',
          staging: `http://${h.svc(services.authAdminApi)}`,
        })),
        prod: json({
          dev: 'https://services-auth-admin-api.internal.dev01.devland.is',
          staging:
            'https://services-auth-admin-api.internal.staging01.devland.is',
          prod: 'https://services-auth-admin-api.internal.innskra.island.is',
        }),
```

File: libs/clients/auth/admin-api/src/lib/auth-admin-api-client.config.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+type BasePaths = z.infer<typeof schema>['basePaths']
+
+export const AuthAdminApiClientConfig = defineConfig({
+  name: 'AuthAdminApiClientConfig',
+  schema,
+  load(env) {
+    return {
+      basePaths: env.requiredJSON<BasePaths>('AUTH_ADMIN_API_PATHS', {
+        dev: 'http://localhost:6333',
+      }),
```
Comment (eirikurn) [inline libs/clients/auth/admin-api/src/lib/auth-admin-api-client.config.ts:22]: Huh. Did you need to specify the generics argument here?

**Review by eirikurn — APPROVED** [review]

File: libs/clients/auth/admin-api/src/lib/auth-admin-api-client.config.ts (hunk 6 above)
Comment (saevarma) [reply libs/clients/auth/admin-api/src/lib/auth-admin-api-client.config.ts]: Nope, TS in webstorm was messing with me.

File: apps/api/infra/api.ts (hunk 5 above)
Comment (saevarma) [reply apps/api/infra/api.ts]: It worked after changing the return type of the `json` function.

**Review by busla — APPROVED** [review]
LGTM

File: apps/services/auth/admin-api/src/environments/environment.ts [hunk 7]
```diff
@@ -1,4 +1,9 @@
 const audience = ['@island.is/auth/admin', '@admin.island.is']
+const issuer = [
+  'https://identity-server.dev01.devland.is',
+  'https://identity-server.staging01.devland.is',
+  'https://innskra.island.is',
+]
```
Comment (eirikurn) [inline apps/services/auth/admin-api/src/environments/environment.ts:6]: We should not allow dev/staging tokens in production 🤓

**Review by eirikurn — APPROVED** [review]

**Pull request merged**